### DISCUSSION TITLE
InvenTree Part Import - A CLI tool to automatically import components from DigiKey, Mouser, LCSC, etc.

### DISCUSSION THREAD

**Discussion by 30350n** [inventree/InvenTree#5904] [Show and tell]
## [InvenTree Part Import](https://github.com/30350n/inventree_part_import)

So for the past couple of weeks I've been working on my own CLI tool to quickly import all the electronic components I've ordered over the years into my InvenTree instance. I just finished implementing all the main features I wanted, so [Version 1 is available now at PyPI](https://pypi.org/project/inventree-part-import/)!

### Features

- easily configurable category/parameter setup: simple yaml file structure, config can be initialized from an existing InvenTree instance
- automatic category and parameter matching via aliases
- concurrent part importing from multiple suppliers with minimal api calls / web scraping (for suppliers which don't offer a proper API)
- DigiKey, LCSC, Mouser, Reichelt, TME support (other suppliers can be added relatively easily)
- configurable python hooks to do stuff like assigning categories by component footprint
- dry-run mode to use the tool without actually committing any changes to your InvenTree database (for testing purposes)

### Motivation

This has been born out of some issues I've had with Ki-nTree (which from what I can tell is what most people use for this atm). The main ones being:
- somewhat limited category setup (when I used it you could only nest categories up to 3 levels deep for some reason)
- no automatic category/parameter matching
- no/limited (i don't fully remember tbh) parameter importing
- general ease of use/import speed (entering mpns one by one just doesn't cut it for me when importing hundreds of parts)

(I also honestly just didn't need the whole KiCad part of it, so yeah.)

So I decided to make my own CLI tool, mainly focusing on quickly importing lots of parts with minimal manual intervention required.

I've already done quite some beta testing with a couple of people, but small issues might still be present, so I'd be very happy about any feedback, potential issues and possible feature ideas. Thanks a lot for reading!

~ Bobbe

**Comment by eeintech** [comment]
@30350n Checkout https://github.com/sparkmicro/Ki-nTree/blob/main/run_tests.py maybe you'll find some inspiration there (or not). It was not meant for importing from a source (like a file or MPN list) but may be helpful as reference for your CLI. Btw I might be interested to integrate it into Ki-nTree too 😄

**Reply by 30350n** [reply]
Oh the CLI Tool with the feature set I described is already fully working, but thanks still! 😅 I've already taken a closer look at your code (mainly as a reference for using the digikey/mouser APIs).

**Reply by eeintech** [reply]
@30350n My bad I missed the link, I gave it a shot. I was originally confused why did you need to manage InvenTree categories, given your tool directed me to the supplier part directly I thought it was only the supplier parts which were created. Then I realized that a part was created too! It matched the category I would have picked but the internal part number format is incorrect. For a single part, I am not quite sure why using CLI vs GUI (eg. Ki-nTree or built-in InvenTree import plugin) with option for editing the info that gets added to the InvenTree database.
That say I could see the benefit for pulling multiple supplier parts in one shot, but then how do a user control the multiple available config options available such as: part ID (if it exists) or name/description/category/IPN, attachments, etc. Do you handle this in Excel/CSV option? How do you proceed (eg. step by step pulling data -> editing spreadsheet ->upload)?

**Reply by 30350n** [reply]
> I thought it was only the supplier parts which were created. Then I realized that a part was created too

Yep, it always creates a Part + Manufacturer Part and then one Supplier Part per supplier that has the part.

> but the internal part number format is incorrect

I don't use/touch internal part numbers so not sure what you mean with that.

> how do a user control the multiple available config options available such as: part ID (if it exists) or name/description/category/IPN, attachments, etc.

There's not much to be controlled there. Name, description, category, parameters, etc. get taken directly from the supplier. The only cases were user intervention is required is when a category or parameter couldn't be matched, in which case the CLI will ask you to select/set it manually.

> Do you handle this in Excel/CSV option?

Yes, this works the same, regardless if parts get imported by MPN directly or if the MPN gets extracted from a CSV file.

**Comment by SchrodingersGat** [comment]
@30350n this looks really nice! We should link to this in our [third party tools section](https://inventree.org/extend/integrate/)

**Reply by 30350n** [reply]
Thanks! Sure, I'd love that :)

**Reply by 30350n** [reply]
Should I create a PR for that?

**Reply by SchrodingersGat** [reply]
@30350n that'd be great - head over to the [website repo](https://github.com/inventree/inventree-website/) for that :)

**Comment by SergeoLacruz** [comment]
Good work. Thanks for that.  We have several tools now that integrate with Mouser and others in one or the other way. It makes visible that everyone uses Inventree in a different way. E.g. we cannot create parts automatically because we need a unique IPN that follows a certain scheme. Others do not care about that. It is interesting to see the different use cases.

**Reply by 30350n** [reply]
Thanks!

> E.g. we cannot create parts automatically because we need a unique IPN that follows a certain scheme

Does that IPN have to be set manually by the user though? Or could this be automated? (If so, feel free to create a feature request for that, should be a quick thing to implement).

**Reply by SergeoLacruz** [reply]
Thanks for the offer. We also have to keep it syncronous with CAD and PLM. Actually it is OK to set it manually. Adding components does not happen so often.
